src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -892,6 +893,136 @@ impl ToQemuParams for Serial {
     }
 }
 
+fn format_fds(files: &[File]) -> String {
+    files
+        .iter()
+        .map(|file| file.as_raw_fd().to_string())
+        .collect::<Vec<String>>()
+        .join(":")
+}
+
+#[derive(Debug)]
+struct Netdev {
+    id: String,
+
+    // File descriptors for vhost multi-queue support.
+    // {
+    //      queue_fds: Vec<File>,
+    //      vhost_fds: Vec<File>,
+    // }
+    fds: HashMap<String, Vec<File>>,
+
+    // disable_vhost_net disables virtio device emulation from the host kernel instead of from qemu.
+    disable_vhost_net: bool,
+}
+
+impl Netdev {
+    fn new(id: &str, host_if_name: &str, num_queues: u32) -> Result<Netdev> {
+        let (tun_files, vhost_files) = generate_netdev_fds(host_if_name, num_queues)?;
+        let fds = HashMap::from([
+            ("fds".to_owned(), tun_files),
+            ("vhostfds".to_owned(), vhost_files),
+        ]);
+        for file in fds.values().flatten() {
+            clear_cloexec(file.as_raw_fd()).context("clearing O_CLOEXEC failed")?;
+        }
+
+        Ok(Netdev {
+            id: id.to_owned(),
+            fds,
+            disable_vhost_net: false,
+        })
+    }
+
+    fn set_disable_vhost_net(&mut self, disable_vhost_net: bool) -> &mut Self {
+        self.disable_vhost_net = disable_vhost_net;
+        self
+    }
+}
+
+#[async_trait]
+impl ToQemuParams for Netdev {
+    async fn qemu_params(&self) -> Result<Vec<String>> {
+        let mut params: Vec<String> = Vec::new();
+        params.push("tap".to_owned());
+        params.push(format!("id={}", self.id));
+
+        if !self.disable_vhost_net {
+            params.push("vhost=on".to_owned());
+            if let Some(vhost_fds) = self.fds.get("vhostfds") {
+                params.push(format!("vhostfds={}", format_fds(vhost_fds)));
+            }
+        }
+
+        if let Some(tuntap_fds) = self.fds.get("fds") {
+            params.push(format!("fds={}", format_fds(tuntap_fds)));
+        }
+
+        Ok(vec!["-netdev".to_owned(), params.join(",")])
+    }
+}
+
+#[derive(Debug)]
+pub struct DeviceVirtioNet {
+    // driver is the qemu device driver
+    device_driver: String,
+
+    // id is the corresponding backend net device identifier.
+    netdev_id: String,
+
+    // mac_address is the guest-side networking device interface MAC address.
+    mac_address: Address,
+
+    // disable_modern prevents qemu from relying on fast MMIO.
+    disable_modern: bool,
+
+    num_queues: u32,
+}
+
+impl DeviceVirtioNet {
+    fn new(netdev_id: &str, mac_address: Address) -> DeviceVirtioNet {
+        DeviceVirtioNet {
+            device_driver: "virtio-net-pci".to_owned(),
```

Comment:
Agreed.  This is unfortunately how much of this kind of work has to be done though, now and in the future, since I can't really develop support for archs I have no access to and thus cannot test.  Exactly as you write, this is partial support for PCI only, to be completed later.

Reply:
FYI: I can see the following log from go-runtime:

```
-device driver=virtio-net-ccw,netdev=network-0,mac=d6:f0:4c:10:8d:65,mq=on,devno=fe.0.0006
```

Reply:
If there isn't an actual scenario that mandates to have a PCI device on the QEMU command line for s390x, I'd rather not support PCI at all there. @BbolroC can hopefully shed some light on what is needed for Z.